File: java/CavanJava/src/com/cavan/java/CavanAccelFreqParser.java
package com.cavan.java;

public class CavanAccelFreqParser {

	private int mFreq;

	private CavanPeakValleyFinder mFinderX;
	private CavanPeakValleyFinder mFinderY;
	private CavanPeakValleyFinder mFinderZ;
	private CavanPeakValleyFinder mFinderBest;

	protected void onFreqChanged(int freq) {}

	public CavanAccelFreqParser(long timeFuzz, double valueFuzz) {
		mFinderX = new CavanPeakValleyFinder(timeFuzz, valueFuzz);
		mFinderY = new CavanPeakValleyFinder(timeFuzz, valueFuzz);
		mFinderZ = new CavanPeakValleyFinder(timeFuzz, valueFuzz);

		mFinderBest = mFinderX;
	}

	public int getFreq() {
		return mFinderBest.getFreq();
	}

	public void setValueFuzz(double fuzz) {
		mFinderX.setValueFuzz(fuzz);
		mFinderY.setValueFuzz(fuzz);
		mFinderZ.setValueFuzz(fuzz);
	}

	public void setTimeFuzz(long fuzz) {
		mFinderX.setTimeFuzz(fuzz);
		mFinderY.setTimeFuzz(fuzz);
		mFinderZ.setTimeFuzz(fuzz);
	}

	public CavanPeakValleyFinder getFinder() {
		return mFinderBest;
	}

	private void updateFreq(int freq) {
		if (freq > 0 && mFreq > 0) {
			freq = (mFreq + freq) / 2;
		}

		if (mFreq != freq) {
			mFreq = freq;
			onFreqChanged(freq);
		}
	}

	public void putValue(double x, double y, double z) {
		mFinderX.putFreqValue(x);
		mFinderY.putFreqValue(y);
		mFinderZ.putFreqValue(z);

		if (mFinderX.getAvgDiff() > mFinderY.getAvgDiff()) {
			if (mFinderX.getAvgDiff() > mFinderZ.getAvgDiff()) {
				mFinderBest = mFinderX;
			} else {
				mFinderBest = mFinderZ;
			}
		} else if (mFinderY.getAvgDiff() > mFinderZ.getAvgDiff()) {
			mFinderBest = mFinderY;
		} else {
			mFinderBest = mFinderZ;
		}

		updateFreq(mFinderBest.getFreq());
	}

	public void putData(CavanAccelSensor sensor) {
		putValue(sensor.getAxisX(), sensor.getAxisY(), sensor.getAxisZ());
	}
}

File: xcode/JwaooToy/JwaooToy/CavanAccelFreqParser.m
//
//  CavanAccelFreqParser.m
//  JwaooToy
//
//  Created by 曹福昂 on 16/7/8.
//  Copyright © 2016年 曹福昂. All rights reserved.
//

#import "CavanAccelFreqParser.h"

#pragma clang diagnostic ignored "-Warc-performSelector-leaks"

@implementation CavanAccelFreqParser

@synthesize freq = mFreq;

- (CavanAccelFreqParser *)initWithValueFuzz:(double)valueFuzz
                          withTimeFuzz:(NSTimeInterval)timeFuzz
{
    if (self = [super init]) {
        mFinderX = [[CavanPeakValleyFinder alloc] initWithValueFuzz:valueFuzz withTimeFuzz:timeFuzz];
        mFinderY = [[CavanPeakValleyFinder alloc] initWithValueFuzz:valueFuzz withTimeFuzz:timeFuzz];
        mFinderZ = [[CavanPeakValleyFinder alloc] initWithValueFuzz:valueFuzz withTimeFuzz:timeFuzz];
        mFinderBest = mFinderX;
    }

    return self;
}

- (void)setFreqSelector:(SEL)selector
             withTarget:(NSObject *)target {
    mFreqSelector = selector;
    mFreqTarget = target;
}

- (void)onFreqChanged:(int)freq {
    [mFreqTarget performSelector:mFreqSelector withObject:[NSNumber numberWithInt:freq]];
}

- (void)updateFreq:(int)freq {
    if (freq > 0 && mFreq > 0) {
        freq = (mFreq + freq) / 2;
    }

    if (mFreq != freq) {
        mFreq = freq;
        [self onFreqChanged:freq];
    }
}

- (void)putSensorData:(nonnull CavanAccelSensor *)sensor {
#if 1
    [mFinderX putFreqValue:sensor.x];
    [mFinderY putFreqValue:sensor.y];
    [mFinderZ putFreqValue:sensor.z];

    if (mFinderX.diff > mFinderY.diff) {
        if (mFinderX.diff > mFinderZ.diff) {
            mFinderBest = mFinderX;
        } else {
            mFinderBest = mFinderZ;
        }
    } else if (mFinderY.diff > mFinderZ.diff) {
        mFinderBest = mFinderY;
    } else {
        mFinderBest = mFinderZ;
    }
#else
    [mFinderZ putFreqValue:sensor.z];
    mFinderBest = mFinderZ;
#endif

    [self updateFreq:mFinderBest.freq];
}

@end
